crates/sound/src/device.rs

```diff
@@ -0,0 +1,717 @@
+                            streams[stream_id as usize].buffers.push_back(b);
+                        }
+                        state = TxState::Done;
+                    }
+                    TxState::Ready
+                        if descriptor.len() as usize != size_of::<VirtioSoundPcmXfer>() =>
+                    {
+                        return Err(Error::UnexpectedDescriptorSize(
+                            size_of::<VirtioSoundPcmXfer>(),
+                            descriptor.len(),
+                        )
+                        .into());
+                    }
+                    TxState::Ready => {
+                        let xfer = desc_chain
+                            .memory()
+                            .read_obj::<VirtioSoundPcmXfer>(descriptor.addr())
+                            .map_err(|_| Error::DescriptorReadFailed)?;
+                        let stream_id: u32 = xfer.stream_id.into();
+                        stream_ids.insert(stream_id);
+
+                        state = TxState::WaitingBufferForStreamId(stream_id);
+                    }
+                    TxState::WaitingBufferForStreamId(stream_id)
+                        if descriptor.len() as usize == size_of::<VirtioSoundPcmXfer>() =>
+                    {
+                        return Err(Error::UnexpectedDescriptorSize(
+                            u32::from(
+                                self.streams.read().unwrap()[stream_id as usize]
+                                    .params
+                                    .buffer_bytes,
+                            ) as usize,
+                            descriptor.len(),
+                        )
+                        .into());
+                    }
+                    TxState::WaitingBufferForStreamId(_stream_id) => {
+                        let mut buf = vec![0; descriptor.len() as usize];
+                        let bytes_read = desc_chain
+                            .memory()
+                            .read(&mut buf, descriptor.addr())
+                            .map_err(|_| Error::DescriptorReadFailed)?;
+                        buf.truncate(bytes_read);
+
+                        buffers.push(Buffer::new(buf, Arc::clone(&message)));
+                    }
+                }
+            }
+        }
+
+        if !stream_ids.is_empty() {
+            let b = audio_backend.write().unwrap();
+            for id in stream_ids {
+                b.write(id).unwrap();
+            }
+        }
+
+        Ok(false)
+    }
+
+    fn process_rx(
+        &self,
+        _vring: &VringRwLock,
+        _audio_backend: &RwLock<Box<dyn AudioBackend + Send + Sync>>,
+    ) -> IoResult<bool> {
+        log::trace!("process_rx");
+        Ok(false)
+    }
+}
+
+pub struct VhostUserSoundBackend {
+    threads: Vec<RwLock<VhostUserSoundThread>>,
+    virtio_cfg: VirtioSoundConfig,
+    exit_event: EventFd,
+    audio_backend: RwLock<Box<dyn AudioBackend + Send + Sync>>,
```

Comment:
Would it make sense to define `audio_backend` at `VhostUserSoundThread` instead? If that can't change over the life of the thread, it may be better to pass it on just once at creation. This would prevent changing the signature of `process_tx()` and `process_ctr()`. Also, shall we ensure internal mutability for the audio_backend? In that case, shall we re-define audio_backend as `Arc<Box<dyn AudioBackend + Send + Sync>>`?